server/src/main/java/org/elasticsearch/action/SingleResultDeduplicator.java

```diff
@@ -45,50 +46,79 @@ public SingleResultDeduplicator(ThreadContext threadContext, Consumer<ActionList
             } else {
                 // already running an execution, queue this one up
-                waitingListeners.add(ContextPreservingActionListener.wrapPreservingContext(listener, threadContext));
+                waitingListOfListenerWithContext.add(
+                    new ListenerWithContext<>(
+                        ContextPreservingActionListener.wrapPreservingContext(listener, threadContext),
+                        // Only the first listener in queue needs the stored context which is used for running executeAction
+                        waitingListOfListenerWithContext.isEmpty() ? threadContext.newStoredContext() : null
+                    )
+                );
                 return;
             }
         }
-        doExecute(listener);
+        doExecute(ContextPreservingActionListener.wrapPreservingContext(listener, threadContext), null);
     }
 
-    private void doExecute(ActionListener<T> listener) {
+    private void doExecute(ActionListener<T> listener, @Nullable ThreadContext.StoredContext storedContext) {
         final ActionListener<T> wrappedListener = ActionListener.runBefore(listener, () -> {
-            final List<ActionListener<T>> listeners;
+            final List<ListenerWithContext<T>> thisBatchOfListenerWithContext;
             synchronized (this) {
-                if (waitingListeners.isEmpty()) {
+                if (waitingListOfListenerWithContext.isEmpty()) {
                     // no listeners were queued up while this execution ran, so we just reset the state to not having a running execution
-                    waitingListeners = null;
+                    waitingListOfListenerWithContext = null;
                     return;
                 } else {
                     // we have queued up listeners, so we create a fresh list for the next execution and execute once to handle the
                     // listeners currently queued up
-                    listeners = waitingListeners;
-                    waitingListeners = new ArrayList<>();
+                    thisBatchOfListenerWithContext = waitingListOfListenerWithContext;
+                    waitingListOfListenerWithContext = new ArrayList<>();
                 }
             }
-            doExecute(new ActionListener<>() {
-                @Override
-                public void onResponse(T response) {
-                    ActionListener.onResponse(listeners, response);
-                }
 
-                @Override
-                public void onFailure(Exception e) {
-                    ActionListener.onFailure(listeners, e);
-                }
-            });
+            // Create a child threadContext so that the parent context remains unchanged when the child execution ends.
+            // This ensures the parent does not see response headers from the child execution.
+            try (var ignore = threadContext.newRestorableContext(false).get()) {
```

Comment:
I think this is equivalent?

```suggestion
            try (var ignore = threadContext.newStoredContext()) {
```

Reply:
Yes. It is equivalent and simpler. It's now updated.